### DISCUSSION TITLE
Ground elevation using terrain tiles

### DISCUSSION THREAD

**Discussion by schedul-xor** [mapsforge/vtm#1039] [Q&A] [answered]
Mapbox supports terrain elevation view, reading DEM tiles in "terrain-rgb" format. It will be nice to import this idea to vtm.
Currently VTM' vector rendering is done by stenciling map directly per each tile, so it can't draw to terrain polygons. Then, how about stenciling to a framebuffer, and use it as a "map texture"? Applying this texture to a terrain-shaped polygons may end up to terrain elevation views.

VTM already supports sun shade, so calculating hillshades to terrain-shaped polygons will surely give us great rendering results.

Reference: https://docs.mapbox.com/mapbox-gl-js/example/query-terrain-elevation/

**Comment by demiantres** [comment]
Do you want to show a 3D terrain, or just hill shading on the 2D map?

**Reply by schedul-xor** [reply]
3D terrains.

**Reply by demiantres** [reply, marked as answer]
For 3D terrains you will have to rewrite the whole VTM engine because everything (POI, lines, ...) is 2D.

**Comment by jhotadhari** [comment]
I asked claude code with `deepseek-v4-pro[1m]` model this:

> this repository has an open discussion about this repo: https://github.com/mapsforge/vtm/discussions/1039
> is it possible to render 3D terrain?
> but not on the base of DEM tiles in "terrain-rgb" format. Its better to use the existing hgt

That is the answer:

## Current State: What VTM Already Has

The claim in the discussion that "you will have to rewrite the whole VTM engine" is **overly pessimistic but not entirely wrong** — it depends on what exactly you mean by "3D terrain." VTM already has significant 3D infrastructure:

### ✅ Already built and working:

| Capability | How it works |
|---|---|
| **Perspective camera** | Full 3D frustum projection with tilt (0–65°), bearing, roll. `Viewport`/`ViewController` builds proper view + projection matrices |
| **3D building extrusion** | `ExtrusionRenderer` + `ExtrusionBucket` render extruded OSM footprints with roofs, walls, lighting, and shadow mapping |
| **Triangle mesh rendering** | `ExtrusionBucket.addMesh()` + `extrusion_layer_mesh.glsl` shader — arbitrary triangle meshes with per-face normals and lighting |
| **3D model overlay** | `vtm-gdx-poi3d` places libGDX 3D models (trees, cars, lamps) geo-referenced on the map with proper depth testing |
| **HGT file reading** | `vtm-hillshading` module reads SRTM `.hgt` files via Mapsforge's `MemoryCachingHgtReaderTileSource` |
| **Depth buffer** | Proper `GL.DEPTH_BUFFER_BIT`, `depthMask`, `depthFunc` throughout the pipeline |
| **2D hillshading** | `HillshadingTileDataSource` generates shaded relief bitmap tiles from HGT data, overlaid as a `BitmapTileLayer` |

The shader at `vtm/resources/assets/shaders/extrusion_layer_mesh.glsl` already supports:
- Per-vertex position with configurable height (`pos.z *= u_alpha`)
- Per-face normals (2-byte encoded, reconstructed in the vertex shader)
- Directional lighting with dot-product diffuse shading
- Shadow mapping via depth texture with PCF filtering
- SSAO approximation based on vertex height

### ❌ What does NOT exist:

- **No 3D terrain mesh** — the map surface is always a flat 2D plane
- **No vector data on terrain** — roads, labels, landuse polygons are all rendered in 2D
- **No terrain LOD system** — no quadtree-based mesh simplification
- **No texture projection** onto a terrain surface

---

## The Real Challenge: Two Different Problems

### Problem 1: 3D Terrain Mesh (Relatively Easy)

Generating a 3D heightfield mesh from HGT files and rendering it is **feasible** with VTM's existing infrastructure:

- You'd create a new `TerrainLayer` (similar to `BuildingLayer`) that fetches HGT data for visible tiles
- Generate a triangle mesh using `ExtrusionBucket.addMesh()` (which already handles vertex buffers, normals, indices)
- Render it with the existing `ExtrusionRenderer` mesh path
- Use VTM's existing perspective projection and depth buffer

This would give you a **bare 3D terrain surface** — mountains and valleys in 3D — but the vector map data (roads, labels, etc.) would still render on the **flat 2D plane underneath or above it**, not draped onto the terrain surface.

### Problem 2: Map Data on Terrain (The Hard Part)

This is what the discussion was really about: rendering vector map data **on top of** a 3D terrain surface. This is genuinely hard because:

1. **VTM renders map data per-tile as 2D raster textures** — `TileLayer` composites vector data onto tile-sized bitmaps.
2. To drape these onto terrain, you'd need to either:
   - **Texture-map approach**: Render each tile's bitmap as a texture onto the corresponding terrain mesh region. This means converting the render-to-texture pipeline to output OpenGL textures instead of just bitmaps for display.
   - **Per-vertex rendering**: Render every road, polygon, and label as 3D geometry that conforms to the terrain height. This is a complete rearchitecture of the rendering engine.
3. The tile pyramid would need to generate terrain meshes at every zoom level, with proper LOD transitions.
4. Labels, icons, and other screen-space elements need special handling to not sink into terrain.

**This is why the discussion concluded it would require a rewrite** — for *textured terrain*, the conclusion is approximately correct. VTM's rendering pipeline is fundamentally 2D-tile-first.

---

## What You *Can* Do With HGT Today

### Option A: Hillshading + 3D Buildings (Already Works)

```java
// Desktop example from MapsforgeTest.java
HillshadingTileSource hillshading = new HillshadingTileSource(
    Viewport.MIN_ZOOM_LEVEL, Viewport.MAX_ZOOM_LEVEL,
    new DemFolderFS(new File("/path/to/hgt/files")),
    new AdaptiveClasyHillShading(),
    128, Color.BLACK, AwtGraphicFactory.INSTANCE);
mMap.layers().add(new BitmapTileLayer(mMap, hillshading, 150));

// Add 3D buildings
mMap.layers().add(new S3DBLayer(mMap, s3dbTileSource));
mMap.layers().add(new BuildingLayer(mMap));
```

Tilt the map and you get a convincing 3D-like view: shaded relief underneath extruded buildings with proper perspective and depth.

### Option B: Bare 3D Terrain Mesh (Feasible, ~1-2 weeks of work)

You could build a `TerrainLayer` that:
1. Reads HGT tiles via Mapsforge's existing `MemoryCachingHgtReaderTileSource`
2. Generates a triangle mesh heightfield
3. Feeds it into `ExtrusionBucket.addMesh()` for rendering
4. Uses the existing mesh shader with normals + lighting

This would give you a gray/untextured 3D terrain surface. Combine it with 3D buildings, and you get something like a basic 3D city viewer. The map data (roads, labels) would render flat, but they'd be hidden by the terrain where it rises.

### Option C: Terrain-RGB Approach (Moderate Effort)

The discussion mentioned Mapbox's terrain-rgb format. You could:
1. Use existing terrain-rgb tile servers (or host your own from HGT → RGB encoding)
2. Create a new `TileSource` that fetches terrain-rgb tiles
3. Generate a terrain mesh from the decoded elevation values per tile
4. Render it as mesh geometry

This is similar to Option B but uses tile-server infrastructure instead of local HGT files.

---

## Bottom Line

| Goal | Feasibility | Effort |
|---|---|---|
| 2D hillshading from HGT | ✅ Already exists | 0 — use `vtm-hillshading` |
| 3D buildings on flat map | ✅ Already exists | 0 — use `S3DBLayer` / `BuildingLayer` |
| 3D terrain mesh (no map texture) | ✅ Feasible | ~1–2 weeks — extend `ExtrusionRenderer` mesh path |
| 3D terrain with draped vector data | ⚠️ Very hard | Months — requires fundamental render-pipeline changes |
| 3D terrain with raster tile texture | ⚠️ Hard | ~1 month — need render-to-texture + texture coordinate mapping |

The gap between "3D terrain mesh" and "3D terrain with map data on it" is enormous. The former is an extension of existing infrastructure; the latter touches every part of the rendering pipeline.
